File: HeraldTest-Flutter/android/app/src/main/java/com/example/flutterxherald/PeerInfo.java
package com.example.flutterxherald;

import java.util.ArrayList;
import io.heraldprox.herald.sensor.datatype.Date;

// https://github.com/broadinstitute/operation-outbreak/issues/163
public class PeerInfo {
    ArrayList<Double> data;
    IllnessStatus status;
    Date lastSeen;
    int updateCount;

    PeerInfo() {
        this.data = new ArrayList<>();
        this.status = new IllnessStatus(IllnessStatusCode.susceptible, new Date());
        this.lastSeen = new Date();
        this.updateCount = 0;
    }

    void setStatus(IllnessStatus status) {
        this.status = status;
        this.lastSeen = new Date();
        this.updateCount += 1;
    }

    void addRSSI(double value) {
        if (-100 <= value && value < 0) {
            data.add(value);
            if (25 < data.size()) {
                data.remove(0);
            }
        }

    }

    double getRSSI() {
        if (0 < data.size()) {
            double sum = 0;
            for (Double v: data) sum += v;
            return sum / data.size();
        } else {
            return -50;
        }
    }
}

File: HeraldTest-Flutter/ios/Runner/PeerInfo.swift
//
//  PeerInfo.swift
//  Runner
//
//  Created by Andi Grozdani on 11/2/21.
//

import Foundation

class PeerInfo {
    public var data: [Double] = []
    public var status: IllnessStatus
    public var lastSeen: Date
    public var updateCount: Int
    
    init(data: [Double], status: IllnessStatus, lastSeen: Date) {
        self.data = data
        self.status = status
        self.lastSeen = lastSeen
        self.updateCount = 0
    }
    
    init() {
        self.data = []
        self.status = IllnessStatus.init(status: .susceptible, dateSince: Date())
        self.lastSeen = Date()
        self.updateCount = 0
    }
    
    func setStatus(_ status: IllnessStatus) {
        self.status = status
        self.lastSeen = Date()
        self.updateCount += 1
    }
    
    func addRSSI(_ value: Double) {
        if (-100 <= value && value < 0) {
            data.append(value);
            if (25 < data.count) {
               data.remove(at:0);
            }
        }
    }
    
    func getRSSI() -> Double {
        if (0 < data.count) {
            var sum = 0.0;
            for v in data {
                sum += v
            }
            return sum / Double(data.count);
        } else {
            return -50;
        }
    }
}
